### ISSUE TITLE
Visualize nuscenes train results

### ISSUE THREAD

**Issue by jing-zhao9** [Pointcept/Pointcept#296]
After running the nuscenes code, I output the labels for each point and store them in the npy file. How can I use this label to visualize the results of nuscenes and compare them with the ground truth?

**Comment by Gofinge** [comment]
You have to write a script to map these predictions into a color and then export them to a ply file. There is some tool in pointcept.utils.visualization.save_point_cloud.

**Comment by jing-zhao9** [comment]
Thank you for your patient answer! The problem I encountered is that the pred.npy in the results folder of the model output is only the label of each point (without coordinate information). However, due to the large number of samples in the Nuscenes dataset. This caused me to be unable to find the original file of the corresponding test set from the output pred.npy. Because it requires the coordinate information of the files in the original test set and the label information of each point in the pred.npy file to be visualized.

**Comment by Gofinge** [comment]
I think the order of points is the same as processed data. You can get the coordinate we dataset by the data name?

**Comment by william-ngvn** [comment]
> Thank you for your patient answer! The problem I encountered is that the pred.npy in the results folder of the model output is only the label of each point (without coordinate information). However, due to the large number of samples in the Nuscenes dataset. This caused me to be unable to find the original file of the corresponding test set from the output pred.npy. Because it requires the coordinate information of the files in the original test set and the label information of each point in the pred.npy file to be visualized.

Hi. Have you solved the problem and visualized it successfully? Many thank you.

**Comment by jacoblambert** [comment]
I changed tester class to also save .pcd file with color, so maybe this helps you. Requires pypcd4 library.
```
class SemSegTester(TesterBase):
    def test(self):
        assert self.test_loader.batch_size == 1
        logger = get_root_logger()
        logger.info(">>>>>>>>>>>>>>>> Start Evaluation >>>>>>>>>>>>>>>>")

        batch_time = AverageMeter()
        intersection_meter = AverageMeter()
        union_meter = AverageMeter()
        target_meter = AverageMeter()
        self.model.eval()

        save_path = os.path.join(self.cfg.save_path, "result")
        make_dirs(save_path)
        # create submit folder only on main process
        if (
            self.cfg.data.test.type == "ScanNetDataset"
            or self.cfg.data.test.type == "ScanNet200Dataset"
            or self.cfg.data.test.type == "ScanNetPPDataset"
        ) and comm.is_main_process():
            make_dirs(os.path.join(save_path, "submit"))
        elif (
            self.cfg.data.test.type == "SemanticKITTIDataset" and comm.is_main_process()
        ):
            make_dirs(os.path.join(save_path, "submit"))
        elif self.cfg.data.test.type == "NuScenesDataset" and comm.is_main_process():
            import json

            make_dirs(os.path.join(save_path, "submit", "lidarseg", "test"))
            make_dirs(os.path.join(save_path, "submit", "test"))
            submission = dict(
                meta=dict(
                    use_camera=False,
                    use_lidar=True,
                    use_radar=False,
                    use_map=False,
                    use_external=False,
                )
            )
            with open(
                os.path.join(save_path, "submit", "test", "submission.json"), "w"
            ) as f:
                json.dump(submission, f, indent=4)
        comm.synchronize()
        record = {}
        # fragment inference
        for idx, data_dict in enumerate(self.test_loader):
            end = time.time()
            data_dict = data_dict[0]  # current assume batch size is 1
            fragment_list = data_dict.pop("fragment_list")
            segment = data_dict.pop("segment")
            data_name = data_dict.pop("name")
            pred_save_path = os.path.join(save_path, "{}_pred.npy".format(data_name))
            pcd_save_path = os.path.join(save_path, "{}_pred.pcd".format(data_name))
            
            if os.path.isfile(pred_save_path):
                logger.info(
                    "{}/{}: {}, loaded pred and label.".format(
                        idx + 1, len(self.test_loader), data_name
                    )
                )
                pred = np.load(pred_save_path)
                if "origin_segment" in data_dict.keys():
                    segment = data_dict["origin_segment"]
            else:
                pred = torch.zeros((segment.size, self.cfg.data.num_classes)).cuda()
                feat = torch.zeros((segment.size, 4)).cuda()
                print(pred.shape, feat.shape)
                for i in range(len(fragment_list)):
                    fragment_batch_size = 1
                    s_i, e_i = i * fragment_batch_size, min(
                        (i + 1) * fragment_batch_size, len(fragment_list)
                    )
                    input_dict = collate_fn(fragment_list[s_i:e_i])
                    for key in input_dict.keys():
                        if isinstance(input_dict[key], torch.Tensor):
                            input_dict[key] = input_dict[key].cuda(non_blocking=True)
                    idx_part = input_dict["index"]
                    with torch.no_grad():
                        pred_part = self.model(input_dict)["seg_logits"]  # (n, k)
                        pred_part = F.softmax(pred_part, -1)
                        if self.cfg.empty_cache:
                            torch.cuda.empty_cache()
                        bs = 0
                        for be in input_dict["offset"]:
                            pred[idx_part[bs:be], :] += pred_part[bs:be]
                            feat[idx_part[bs:be], :] = input_dict["feat"][bs:be]
                            print(f"intermediate pred shape: {pred.shape}, feat shape: {feat.shape}")
                            bs = be

                    logger.info(
                        "Test: {}/{}-{data_name}, Batch: {batch_idx}/{batch_num}".format(
                            idx + 1,
                            len(self.test_loader),
                            data_name=data_name,
                            batch_idx=i,
                            batch_num=len(fragment_list),
                        )
                    )
                if self.cfg.data.test.type == "ScanNetPPDataset":
                    pred = pred.topk(3, dim=1)[1].data.cpu().numpy()
                else:
                    pred = pred.max(1)[1].data.cpu().numpy()
                if "origin_segment" in data_dict.keys():
                    assert "inverse" in data_dict.keys()
                    print(pred.shape)
                    pred = pred[data_dict["inverse"]]
                    print(pred.shape)
                    segment = data_dict["origin_segment"]
                    feat = feat[data_dict["inverse"]]
                print("pred shape: {}, segment shape: {}, feat shape: {}".format(np.array(pred).shape, np.array(segment).shape, feat.shape))
                np.save(pred_save_path, pred)
                xyzil = np.hstack((feat.cpu(), pred.reshape(-1,1)))
                pypcd4.PointCloud.from_xyzil_points(xyzil).save(pcd_save_path)
```

But result is disappointing:

![Image](https://github.com/user-attachments/assets/b68a69f2-03d2-4892-9fc8-f40489b20121)

Any ideas why inference result is so poor? In general I got really low mAP. The ground seems OK so I don't think it's issue with pointcloud assembly, but it's possible.

**Comment by jacoblambert** [comment]
I learned from various PTv3 issues that I had to use specifically branch v1.5.1 and install specific OctFormer dependencies, https://github.com/Pointcept/Pointcept?tab=readme-ov-file#octformer 

Then it works well with the above visualization

![Image](https://github.com/user-attachments/assets/5708f350-0645-41e5-a8ce-250831a6375c)

**Comment by why201520** [comment]
I used your code, and the resulting y and z coordinates are both 0. May I ask why this is happening?

**Comment by why201520** [comment]
There's a line of code that reads: feat[idx_part[bs:be], :] = input_dict["feat"][bs:be]. Shouldn't this actually be feat[idx_part[bs:be], :] = input_dict["coord"][bs:be]? Because feat is supposed to store the point cloud's features, while coord represents the 3D coordinates of the points, right?"

**Comment by Gofinge** [comment]
Hey all, hope our recently demo code for outdoor point cloud is helpful: https://github.com/Pointcept/Concerto/blob/main/demo/0_pca_outdoor.py
